# Daria Bromot (FS500047)

## Online Retail Exercise

**Import pandas and read in the Online Retail csv file and set it to a DataFrame called retail**

In [15]:
import pandas as pd

In [16]:
retail = pd.read_csv('Online Retail.csv', encoding='latin-1')

**Check the head of the DataFrame.**

In [17]:
retail.head(10)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

        InvoiceDate  UnitPrice  CustomerID         Country  
0  01/12/2010 08:26       2.55     17850.0  United Kingdom  
1  01/12/2010 08:26       3.39     17850.0  United Kingdom  
2  01/12/2010 08:26       2.75     17850.0  United Kingdom  
3  01/12/2010 08:26       3.39     17850.0  United Kingdom  
4  01/12/2010 08:26       3.39     17850.0  United Kingdom  
5  01/12/2010 08:26       7.65     17850.0  United Kingdom  
6  01/12/2010 08:26       4.25     17850.0  United Kingdom  
7  01/12/2010 08:28       1.85     17850.0  United Kingdom  
8  01/12/2010 08:28       1.85     17850.0  United Kingdom  
9  01/12/2010 08:34       1.69     13047.0  United Kingdom

**Use the .info() method to find out how many entries there are.**

In [18]:
retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


**Notice that the CustomerID column has only 406829 entries while other columns have much more.  Use the .isnull() and .sum() to find the total number of missing values in the dataframe.**

In [19]:
missing_values = retail.isnull().sum()
missing_values 

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [20]:
missing_values.sum()

136534

**What are the highest and lowest unit prices?**

In [21]:
retail['UnitPrice'].max()

38970.0

In [22]:
retail['UnitPrice'].min()

-11062.06

**From which countries were purchases made? What are the top 3 purchasing countries? What are the bottom 3?** 

In [23]:
retail['Country'].value_counts()

United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58
Lebanon 

In [24]:
retail['Country'].value_counts().head(3)

United Kingdom    495478
Germany             9495
France              8557
Name: Country, dtype: int64

In [25]:
retail['Country'].value_counts().tail(3)

Czech Republic    30
Bahrain           19
Saudi Arabia      10
Name: Country, dtype: int64

**How many unique products were purchased?**

In [26]:
retail['Description'].nunique()

4223

**How many people purchased more than 10 of a single item and are from the United Kingdom?**

In [27]:
retail[(retail['Country']=='United Kingdom') & 
       (retail['Quantity']>10)].count()

InvoiceNo      109870
StockCode      109870
Description    109627
Quantity       109870
InvoiceDate    109870
UnitPrice      109870
CustomerID     104109
Country        109870
dtype: int64

**We would now like to determine the percentage of the total invoice that each StockCode represents.  We will do this in two ways: using merging and using transform:**

### Option №1

**First, create a new column with the total price for each item purchased (based on quantity and unit price)**

In [38]:
retail['Total Price'] = retail['Quantity'] * retail['UnitPrice']

In [39]:
retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 11 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   InvoiceNo            541909 non-null  object 
 1   StockCode            541909 non-null  object 
 2   Description          540455 non-null  object 
 3   Quantity             541909 non-null  int64  
 4   InvoiceDate          541909 non-null  object 
 5   UnitPrice            541909 non-null  float64
 6   CustomerID           406829 non-null  float64
 7   Country              541909 non-null  object 
 8   Total Price          541909 non-null  float64
 9   Order Total          541909 non-null  float64
 10  Percentage of Order  539794 non-null  float64
dtypes: float64(5), int64(1), object(5)
memory usage: 45.5+ MB


**Determine the total by invoice using grouby**

In [40]:
retail.groupby('InvoiceNo')["Total Price"].sum()

InvoiceNo
536365        139.12
536366         22.20
536367        278.73
536368         70.05
536369         17.85
             ...    
C581484   -168469.60
C581490       -32.53
C581499      -224.69
C581568       -54.75
C581569        -7.50
Name: Total Price, Length: 25900, dtype: float64

**Reset the index of the new data and use merge to combine this new data with the original data.  Add a new column with the percentage of the order named Percentage of Order.**

In [55]:
invoice = retail.groupby('InvoiceNo')["Total Price"]
invoice_total = invoice.sum().rename("Order Total").reset_index()

In [56]:
re_retail = retail.merge(invoice_total)

In [57]:
re_retail["Perentage of Order"] = new_retail["Total Price"] / new_retail["Order Total"]

In [58]:
re_retail.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country  Total Price  \
0  01/12/2010 08:26       2.55     17850.0  United Kingdom        15.30   
1  01/12/2010 08:26       3.39     17850.0  United Kingdom        20.34   
2  01/12/2010 08:26       2.75     17850.0  United Kingdom        22.00   
3  01/12/2010 08:26       3.39     17850.0  United Kingdom        20.34   
4  01/12/2010 08:26       3.39     17850.0  United Kingdom        20.34   

   Order Total  Percentage of Order  Perentage of Order  
0       139.12             0.109977            0.109977  
1       139.12             0.146205            0.146205  
2       139.12             0.158137            0.158137  
3       139.12             0.146205            0.146205  
4       139.12             0.146205            0.146205

### Option №2

**The second approach is to use groupby along with transform.  This avoids the issue of reordering the index to obtain the new dataframe.  We retain the same number of items as in the original data set**

In [51]:
retail.groupby('InvoiceNo')["Total Price"].transform('sum')

0         139.12
1         139.12
2         139.12
3         139.12
4         139.12
           ...  
541904    249.45
541905    249.45
541906    249.45
541907    249.45
541908    249.45
Name: Total Price, Length: 541909, dtype: float64

**Add a column to calculate the percentage of the order.**

In [52]:
retail["Order Total"]=retail.groupby('InvoiceNo')["Total Price"].transform('sum')

In [59]:
retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 11 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   InvoiceNo            541909 non-null  object 
 1   StockCode            541909 non-null  object 
 2   Description          540455 non-null  object 
 3   Quantity             541909 non-null  int64  
 4   InvoiceDate          541909 non-null  object 
 5   UnitPrice            541909 non-null  float64
 6   CustomerID           406829 non-null  float64
 7   Country              541909 non-null  object 
 8   Total Price          541909 non-null  float64
 9   Order Total          541909 non-null  float64
 10  Percentage of Order  539794 non-null  float64
dtypes: float64(5), int64(1), object(5)
memory usage: 45.5+ MB


In [53]:
retail["Percentage of Order"] = retail["Total Price"] / retail["Order Total"]

In [54]:
retail.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country  Total Price  \
0  01/12/2010 08:26       2.55     17850.0  United Kingdom        15.30   
1  01/12/2010 08:26       3.39     17850.0  United Kingdom        20.34   
2  01/12/2010 08:26       2.75     17850.0  United Kingdom        22.00   
3  01/12/2010 08:26       3.39     17850.0  United Kingdom        20.34   
4  01/12/2010 08:26       3.39     17850.0  United Kingdom        20.34   

   Order Total  Percentage of Order  
0       139.12             0.109977  
1       139.12             0.146205  
2       139.12             0.158137  
3       139.12             0.146205  
4       139.12             0.146205